### PR TITLE
Feature/sdk configuration service

### PR DESCRIPTION
**Pull request by nulrich**
# Problem

We want to use settings exposed on the SDK Configuration Service to control features of the SDK.

# Solution

Add SdkConfigurationServiceProvider at root of the application. It will call `/v3/sdk/configurations` API to retrieve the settings. Settings will be available through useSdkConfigurationService.

## Checklist

_put `n/a` if item is not relevant to PR changes_

- [x] Has the CHANGELOG been updated?
- [ ] Has the README been updated?
- [ ] Has the CONTRIBUTING doc been updated?
- [ ] Has the RELEASE_GUIDELINES been updated?
- [ ] Has the TESTING_GUIDELINES been updated?
- [ ] Has the MIGRATION doc been updated for any MAJOR breaking changes?
- [ ] Has the MIGRATION doc been updated for any MINOR breaking changes, including any translation strings or keys changes?
- [ ] Have any new automated tests been implemented or the existing ones changed?
- [ ] Have any new manual tests been written down or the existing ones changed?
- [ ] Have any new strings been translated or the existing ones changed?

### CODE REVIEW

File: src/contexts/useSdkConfigurationService.tsx
Comment (ethan605) [inline src/contexts/useSdkConfigurationService.tsx:48]: Ditto here (`configuration` will have a default value)

**Review by ethan605 — CHANGES_REQUESTED** [review]

File: src/contexts/useSdkConfigurationService.tsx
Comment (nulrich) [reply src/contexts/useSdkConfigurationService.tsx:15]: In that case, I need introduce a 'loading' state to prevent the UI to be displayed until we get the settings from the API.

Comment (ethan605) [reply src/contexts/useSdkConfigurationService.tsx:15]: Yeah sounds reasonable. With poor network conditions, this request can take a significant amount of time

**Review by stefaniacardenas — CHANGES_REQUESTED** [review]
Some comments/questions:
* Loading screen should be displayed during request
* If the feature is not ready, for example, in case we need to wait for the v4 endpoint to be available, I would suggest putting this feature behind a webpack configuration flag, so that it only gets used if the configuration is enabled. Otherwise, hold off merging to development until ready
* Has the cross-device flow been handled?
* Is the feature always enabled? Can it be disabled on development?

File: src/contexts/useSdkConfigurationService.tsx
Comment (nulrich) [reply src/contexts/useSdkConfigurationService.tsx:15]: I'm putting the Spinner (with autoFocusOnInitialScreenTitle support) used in the other screens instead of a blank page..

File: src/components/utils/onfidoApi.ts
Comment (ethan605) [inline src/components/utils/onfidoApi.ts:396]: We don't need to normalise from `onfido_web_sdk` to `onfido-web-sdk`. Our data pipeline has been living with the former value for ages 😅

File: src/contexts/useSdkConfigurationService.tsx
Comment (ethan605) [inline src/contexts/useSdkConfigurationService.tsx:57]: Just the matter of personal taste:
```suggestion
  if (!configuration) {
    return <Fragment>{fallback}</Fragment>
  }
  
  return (
    <SdkConfigurationServiceContext.Provider value={configuration}>
      {children}
    </SdkConfigurationServiceContext.Provider>
  )
}
```

File: webpack.config.babel.js
Comment (ethan605) [inline webpack.config.babel.js:179]: Please backfill this with default configuration of disabled multi-frame capture 🙏

File: webpack.config.babel.js
Comment (ethan605) [inline webpack.config.babel.js:280]: Might be a change by accident 😄

File: src/contexts/useSdkConfigurationService.tsx
Comment (ethan605) [inline src/contexts/useSdkConfigurationService.tsx:38]: Humm why?

File: webpack.config.babel.js
Comment (nulrich) [reply webpack.config.babel.js:280]: 😅

File: webpack.config.babel.js
Comment (nulrich) [reply webpack.config.babel.js:179]: It's planned in the MFC PR, but I can do it now.

Comment (ethan605) [reply webpack.config.babel.js:179]: Ah okay then. You can keep this for a minimal diffs 🙏

File: src/components/utils/onfidoApi.ts
Comment (nulrich) [reply src/components/utils/onfidoApi.ts:396]: So we need to update the Configuration Service which expects "onfido-web-sdk". Or I put back the hardcoded value.

Comment (ethan605) [reply src/components/utils/onfidoApi.ts:396]: You should keep the `SDK_SOURCE` env var and update sdk-configuration-service to accept `onfido_web_sdk`

File: src/contexts/useSdkConfigurationService.tsx
Comment (ethan605) [reply src/contexts/useSdkConfigurationService.tsx:38]: Oh no, I meant why only for `development`?

Comment (nulrich) [reply src/contexts/useSdkConfigurationService.tsx:38]: * The default configuration will be defined in this file. It's the fallback configuration.
* In the webpackconf, it's only a way for devs to override a configuration value without touching the default value.

File: src/components/utils/onfidoApi.ts
Comment (nulrich) [reply src/components/utils/onfidoApi.ts:396]: 👍

File: src/contexts/useSdkConfigurationService.tsx
Comment (ethan605) [reply src/contexts/useSdkConfigurationService.tsx:38]: > In the webpackconf, it's only a way for devs to override a configuration value without touching the default value.

Well I guess we can simplify the _default_ `SDK_CONFIGURATION` at the front-end. This default value will be the fallback in case we are unable to load the remote configuration. So in each env, we'll expect a different value of this. But for the default value, we'll make it ... default, like no MFC, etc.

So IMO we'll be fine to have `SDK_CONFIGURATION` for `PROD_CONFIG` and propagate it to other envs

Comment (nulrich) [reply src/contexts/useSdkConfigurationService.tsx:38]: 👍

File: src/contexts/useSdkConfigurationService.tsx
Comment (stefaniacardenas) [inline src/contexts/useSdkConfigurationService.tsx:15]: Why is this an env variable and it's an empty object? Is this to allow us to pass the configurations we want at development time?

Comment (stefaniacardenas) [reply src/contexts/useSdkConfigurationService.tsx:15]: These could only be changed at build time, though. Do we want that?

Comment (nulrich) [reply src/contexts/useSdkConfigurationService.tsx:15]: * Default configuration is the fallback when we can't reach the server.
* envConfiguration, configuration in webpack, is the specific configuration per env, to override the server configuration (Ex: We may want to disable something specific in E2E).

Comment (stefaniacardenas) [reply src/contexts/useSdkConfigurationService.tsx:15]: I see. Wouldn't it make more sense to expose a function to update the configuration at runtime for a specific env so you don't need to rebuild?

File: src/components/utils/sdkBackend.ts
Comment (nulrich) [reply src/components/utils/sdkBackend.ts]: 👍

File: src/index.tsx
Comment (stefaniacardenas) [inline src/index.tsx:122]: This would allow anyone to override the configurations. I think we should only allow this if the environment is `test` (so for surge demo app and for E2E tests)

**Review by stefaniacardenas — CHANGES_REQUESTED** [review]

File: src/index.tsx
Comment (stefaniacardenas) [inline src/index.tsx:158]: I think we should update the setOptions function too

**Review by stefaniacardenas — CHANGES_REQUESTED** [review]

**Review by stefaniacardenas — COMMENTED** [review]
The PR looks good to me. Are we going to change the endpoint to point at v4 before merging?

**Comment by nulrich** [comment]
The endpoint v4 is not ready, and I would like this code to finish MFC.
But I'll update it to v4 as soon as it's available.

**Comment by nulrich** [comment]
Finally, the v4 is not ready. They advised me to stick on the v3.

**Review by stefaniacardenas — APPROVED** [review]

**Review by stefaniacardenas — APPROVED** [review]

**Pull request merged**
